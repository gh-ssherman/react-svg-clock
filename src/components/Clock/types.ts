export interface ClockProps {
  date: Date;
}
