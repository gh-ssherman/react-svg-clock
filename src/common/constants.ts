export const colors = {
  case: '#3d4849',
  digit: {
    off: '#222224',
    on: '#00dc55',
  },
  face: '#000',
  feet: '#000',
};

export const transition = '0.33s ease-in-out';
