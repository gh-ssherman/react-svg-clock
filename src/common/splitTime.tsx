const pad = (num: number) => num.toString().padStart(2, '0');

/**
 * Split a date into an array of individual 12 hour format time digits.
 * Values from the date input, other than hh:mm:ss, are ignored.
 *
 * @param dt: JS Date. i.e. new Date() or new Date(1999, 12, 31, 2, 30, 45).
 * @returns Array of digits [h, h, m, m, s, s]. i.e. [0, 2, 3, 0, 4, 5].
 */
const splitTime = (dt: Date): number[] => {
  const time =
    pad(dt.getHours() % 12 || 12) + pad(dt.getMinutes()) + pad(dt.getSeconds());

  return time.split('').map((v) => parseInt(v, 10));
};

export default splitTime;
