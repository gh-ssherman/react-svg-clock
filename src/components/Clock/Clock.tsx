import React from 'react';

import Case from './subcomponents/Case';
import Display from './subcomponents/Display';

import { ClockProps } from './types';

/**
 * Clock component which displays the current time via simulated 7-segment
 * LED digits (plus an AM/PM indicator).
 *
 * @param date: JS Date
 * @returns Clock rendered as a JSX Element
 */
const Clock: React.FC<ClockProps> = ({ date }): JSX.Element => {
  const width = 400;
  const height = width / 4;
  const centerH = -width / 2;
  const centerV = -height / 2;

  return (
    <svg
      width='100%'
      viewBox={`${centerH} ${centerV} ${width} ${height}`}
      preserveAspectRatio='xMinYMin meet'
    >
      <g>
        <Case
          width={width}
          height={height}
        />
        <Display
          date={date}
          width={width}
          height={height}
        />
      </g>
    </svg>
  );
};

export default Clock;
