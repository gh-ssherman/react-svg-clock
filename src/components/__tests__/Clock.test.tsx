import { render, waitFor } from '@testing-library/react';
import { colors } from '../../common/constants';

import { Clock } from '../Clock';

describe('Test <Clock /> component', () => {
  test('Should render correctly', async () => {
    const { container } = render(
      <svg>
        <Clock date={new Date(1999, 12, 31, 12, 30, 30)} />
      </svg>,
    );

    const pm = container?.querySelector("text[id^='pm']");
    await waitFor(() => {
      expect(pm).toHaveAttribute(
        'fill',
        colors.digit.on,
      );

      expect(container).toMatchSnapshot();
    });
  });
});
