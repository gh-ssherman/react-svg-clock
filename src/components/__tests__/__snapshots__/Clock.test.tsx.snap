// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Clock /> component Should render correctly 1`] = `
<div>
  <svg>
    <svg
      preserveAspectRatio="xMinYMin meet"
      viewBox="-200 -50 400 100"
      width="100%"
    >
      <g>
        <svg
          x="-200"
          y="-50"
        >
          <defs>
            <radialgradient
              cx="50%"
              cy="0%"
              fx="50%"
              fy="0%"
              id="reflective-glare"
              r="50%"
            >
              <stop
                offset="0"
                stop-color="#000"
                stop-opacity="0.35"
              />
              <stop
                offset="1"
                stop-color="#000"
                stop-opacity="1"
              />
            </radialgradient>
          </defs>
          <g>
            <svg>
              <defs>
                <clippath
                  id="foot-0-clip"
                >
                  <rect
                    height="7.272727272727273"
                    width="36.36363636363637"
                    x="61.81818181818181"
                    y="90.9090909090909"
                  />
                </clippath>
              </defs>
              <circle
                clip-path="url(#foot-0-clip)"
                cx="80"
                cy="84"
                fill="#000"
                r="18.181818181818183"
              />
            </svg>
            <svg>
              <defs>
                <clippath
                  id="foot-1-clip"
                >
                  <rect
                    height="7.272727272727273"
                    width="36.36363636363637"
                    x="301.8181818181818"
                    y="90.9090909090909"
                  />
                </clippath>
              </defs>
              <circle
                clip-path="url(#foot-1-clip)"
                cx="320"
                cy="84"
                fill="#000"
                r="18.181818181818183"
              />
            </svg>
            <rect
              fill="url(#reflective-glare)"
              height="84"
              rx="23.995200959808034"
              ry="50"
              stroke="#3d4849"
              stroke-width="5.99970001499925"
              width="392"
              x="4"
              y="4"
            />
          </g>
        </svg>
        <svg
          x="-200"
          y="-50"
        >
          <g>
            <svg
              height="60"
              viewBox="-1 -1 12 20"
              x="-140"
              y="16"
            >
              <g>
                <polygon
                  fill="#222224"
                  id="segment-a"
                  points="1, 1  2, 0  8, 0  9, 1  8, 2  2, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-b"
                  points="9, 1 10, 2 10, 8  9, 9  8, 8  8, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-c"
                  points="9, 9 10,10 10,16  9,17  8,16  8,10"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-d"
                  points="9,17  8,18  2,18  1,17  2,16  8,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-e"
                  points="1,17  0,16  0,10  1, 9  2,10  2,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-f"
                  points="1, 9  0, 8  0, 2  1, 1  2, 2  2, 8"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-g"
                  points="1, 9  2, 8  8, 8  9, 9  8,10  2,10"
                  style="transition: 0.33s ease-in-out;"
                />
              </g>
            </svg>
            <svg
              height="60"
              viewBox="-1 -1 12 20"
              x="-100"
              y="16"
            >
              <g>
                <polygon
                  fill="#00dc55"
                  id="segment-a"
                  points="1, 1  2, 0  8, 0  9, 1  8, 2  2, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-b"
                  points="9, 1 10, 2 10, 8  9, 9  8, 8  8, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-c"
                  points="9, 9 10,10 10,16  9,17  8,16  8,10"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-d"
                  points="9,17  8,18  2,18  1,17  2,16  8,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-e"
                  points="1,17  0,16  0,10  1, 9  2,10  2,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-f"
                  points="1, 9  0, 8  0, 2  1, 1  2, 2  2, 8"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-g"
                  points="1, 9  2, 8  8, 8  9, 9  8,10  2,10"
                  style="transition: 0.33s ease-in-out;"
                />
              </g>
            </svg>
            <svg
              height="60"
              viewBox="-1 -1 12 20"
              x="-40"
              y="16"
            >
              <g>
                <polygon
                  fill="#00dc55"
                  id="segment-a"
                  points="1, 1  2, 0  8, 0  9, 1  8, 2  2, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-b"
                  points="9, 1 10, 2 10, 8  9, 9  8, 8  8, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-c"
                  points="9, 9 10,10 10,16  9,17  8,16  8,10"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-d"
                  points="9,17  8,18  2,18  1,17  2,16  8,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-e"
                  points="1,17  0,16  0,10  1, 9  2,10  2,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-f"
                  points="1, 9  0, 8  0, 2  1, 1  2, 2  2, 8"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-g"
                  points="1, 9  2, 8  8, 8  9, 9  8,10  2,10"
                  style="transition: 0.33s ease-in-out;"
                />
              </g>
            </svg>
            <svg
              height="60"
              viewBox="-1 -1 12 20"
              x="0"
              y="16"
            >
              <g>
                <polygon
                  fill="#00dc55"
                  id="segment-a"
                  points="1, 1  2, 0  8, 0  9, 1  8, 2  2, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-b"
                  points="9, 1 10, 2 10, 8  9, 9  8, 8  8, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-c"
                  points="9, 9 10,10 10,16  9,17  8,16  8,10"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-d"
                  points="9,17  8,18  2,18  1,17  2,16  8,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-e"
                  points="1,17  0,16  0,10  1, 9  2,10  2,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-f"
                  points="1, 9  0, 8  0, 2  1, 1  2, 2  2, 8"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-g"
                  points="1, 9  2, 8  8, 8  9, 9  8,10  2,10"
                  style="transition: 0.33s ease-in-out;"
                />
              </g>
            </svg>
            <svg
              height="60"
              viewBox="-1 -1 12 20"
              x="60"
              y="16"
            >
              <g>
                <polygon
                  fill="#00dc55"
                  id="segment-a"
                  points="1, 1  2, 0  8, 0  9, 1  8, 2  2, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-b"
                  points="9, 1 10, 2 10, 8  9, 9  8, 8  8, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-c"
                  points="9, 9 10,10 10,16  9,17  8,16  8,10"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-d"
                  points="9,17  8,18  2,18  1,17  2,16  8,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-e"
                  points="1,17  0,16  0,10  1, 9  2,10  2,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-f"
                  points="1, 9  0, 8  0, 2  1, 1  2, 2  2, 8"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-g"
                  points="1, 9  2, 8  8, 8  9, 9  8,10  2,10"
                  style="transition: 0.33s ease-in-out;"
                />
              </g>
            </svg>
            <svg
              height="60"
              viewBox="-1 -1 12 20"
              x="100"
              y="16"
            >
              <g>
                <polygon
                  fill="#00dc55"
                  id="segment-a"
                  points="1, 1  2, 0  8, 0  9, 1  8, 2  2, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-b"
                  points="9, 1 10, 2 10, 8  9, 9  8, 8  8, 2"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-c"
                  points="9, 9 10,10 10,16  9,17  8,16  8,10"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-d"
                  points="9,17  8,18  2,18  1,17  2,16  8,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-e"
                  points="1,17  0,16  0,10  1, 9  2,10  2,16"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#00dc55"
                  id="segment-f"
                  points="1, 9  0, 8  0, 2  1, 1  2, 2  2, 8"
                  style="transition: 0.33s ease-in-out;"
                />
                <polygon
                  fill="#222224"
                  id="segment-g"
                  points="1, 9  2, 8  8, 8  9, 9  8,10  2,10"
                  style="transition: 0.33s ease-in-out;"
                />
              </g>
            </svg>
            <g>
              <circle
                cx="130"
                cy="38.5"
                fill="#00dc55"
                r="4"
              />
              <circle
                cx="130"
                cy="53.5"
                fill="#00dc55"
                r="4"
              />
            </g>
            <g>
              <circle
                cx="229.99999999999997"
                cy="38.5"
                fill="#00dc55"
                r="4"
              />
              <circle
                cx="229.99999999999997"
                cy="53.5"
                fill="#00dc55"
                r="4"
              />
            </g>
            <g>
              <text
                fill="#222224"
                font-size="10"
                id="am"
                style="transition: 0.33s ease-in-out;"
                x="336"
                y="26"
              >
                AM
              </text>
              <text
                fill="#00dc55"
                font-size="10"
                id="pm"
                style="transition: 0.33s ease-in-out;"
                x="336"
                y="38"
              >
                PM
              </text>
            </g>
          </g>
        </svg>
      </g>
    </svg>
  </svg>
</div>
`;
